### ISSUE TITLE
New Rule: Object.freeze module constants that are objects or arrays

### ISSUE THREAD

**Issue by snewcomer** [eslint/eslint#16552]
### Rule details

detect top level objects or arrays and fix them with Object.freeze

### Related ECMAScript feature

No

### What type of rule is this?

Warns about a potential problem

### Example code

```js
❌ Invalid `["error"]`

const DEFAULT_OPTIONS = {
  from: 'email@email.com',
  to: 'email@email.com',
}

class User {}

✅ Valid

const DEFAULT_OPTIONS = Object.freeze({
  from: 'email@email.com',
  to: 'email@email.com',
});

class User {}
```

### Why should this rule be in the core instead of a plugin?

This sort of problem can be especially bug prone for apps that might mutate the top level constant.  My assumption is there are very few, if any, cases you want a mutable top level object or array.

```
const DEFAULT_OPTIONS = {
  from: 'email@email.com',
  to: 'email@email.com',
}

function mutateIt(args) {
  lodash.merge(DEFAULT_OPTIONS, args);
}

function mutateItAgain(newFrom) {
  DEFAULT_OPTIONS.from = newFrom;
}
```

### Participation

- [X] I am willing to submit a pull request to implement this rule.

### Additional comments

A few considerations...

- `Object.freeze` does not error nested mutations.  How do we handle something like this?  Recursive application?  But then what type of performance characteristics does this have.

- Wrapping in Object.freeze gets you the type.  Just a minor interesting note.

```
Readonly<{
    value: 1;
}>
```

- TypeScript `readonly` doesn't reflect on the built JS bundle.  So not a viable solution.

**Comment by mdjermanovic** [comment]
Hi @snewcomer, thanks for the issue!

This rule doesn't relate to new ECMAScript features that reached stage 4 in the preceding 12 months, so per our [rules policies](https://eslint.org/blog/2020/05/changes-to-rules-policies/) it doesn't qualify for inclusion in the core. You can always make a custom rule and publish a plugin.
